composer/trainer/trainer.py

```diff
@@ -140,6 +246,140 @@ def _unpack_evaluators(
     return evaluators
 
 
+def _unpack_grad_accum(grad_accum: Union[int, str], device: Device) -> Tuple[int, bool]:
+    # Set initial grad_accum to 1 if using adaptive
+    adaptive_grad_accum = grad_accum == "auto"
+    if adaptive_grad_accum:
+        grad_accum = 1
+        warnings.warn(textwrap.dedent("""Setting `grad_accum='auto'` is an experimental feature
+            which may cause uncaught Cuda Out of Memory errors. In this case, please manually
+            set grad_accum explicitly to an integer instead.
+            """),
+                      category=UserWarning)
+    # Cannot use adaptive grad accum on CPU
+    if isinstance(device, DeviceCPU) and adaptive_grad_accum:
+        raise ValueError("Cannot use adaptive grad_accum on CPU. Please set grad_accum >= 1")
+    # grad_accum should be int as we've already handled "auto" case
+    if isinstance(grad_accum, str):
+        raise ValueError("grad_accum must be an int or ``auto``")
+    return grad_accum, adaptive_grad_accum
+
+
+def _is_cuda_oom(e: RuntimeError):
+    """Determines if error is CUDA Out of Memory and if adaptive_grad_accum is enabled."""
+    return "CUDA out of memory" in str(e)
+
+
+def _unpack_device(device: Optional[Union[str, Device]], deepspeed_enabled: bool):
+    if not device:
+        device = DeviceGPU() if torch.cuda.is_available() else DeviceCPU()
+    elif isinstance(device, str):
+        if device.lower() == 'cpu':
+            device = DeviceCPU()
+        elif device.lower() == 'gpu':
+            device = DeviceGPU()
+        else:
+            raise ValueError(f'device ({device}) must be one of (cpu, gpu).')
+    return device
+
+
+def _unpack_seed(seed: Optional[int], device: Device):
+    if not seed:
+        seed = reproducibility.get_random_seed()
+
+    # Ensure that each process has a seed = rank_zero_seed + global_rank
+    # This "deterministically different" seed behavior is required to be able
+    # to restore seeds when resuming form checkpoints, since only the
+    # `rank_zero_seed` is stored on state.
+    if seed < 0 or seed > reproducibility.MAX_SEED:
+        raise ValueError(f"Invalid seed: {seed}. It must be on [0; 2**32 - 1)")
+
+    # using int64 to prevent overflow
+    rank_zero_seed = device.tensor_to_device(torch.tensor([seed], dtype=torch.int64))
+    dist.broadcast(rank_zero_seed, src=0)
+    rank_zero_seed = rank_zero_seed.item()
+    assert isinstance(rank_zero_seed, int)
+    seed = rank_zero_seed + dist.get_global_rank()
+    log.info(f"Setting seed to {seed}")
+    return rank_zero_seed, seed
+
+
+def _unpack_max_duration(max_duration: Union[int, str, Time]):
+    if isinstance(max_duration, str):
+        max_duration = Time.from_timestring(max_duration)
+    elif isinstance(max_duration, int):
+        max_duration = Time.from_epoch(max_duration)
```

Comment:
rename to `parse_max_duration`.

question: should this functionality be in `Time`?

Reply:
Makes sense. I can fix by either allowing a timestring as a single-argument in the `Time` constructor, or move this functionality to an `ensure_time` helper in the Time class.

Reply:
Added an `ensure_time` helper function; removed `_unpack_max_duration`.